### ISSUE TITLE
WooCommerce Blocks actions are fired in a different order depending if it is loaded from Core vs Feature plugin

### ISSUE THREAD

**Issue by opr** [woocommerce/woocommerce-gutenberg-products-block#7294]
## Describe the bug
The order in which actions are fired is different depending where WooCommerce Blocks is loaded from. The two main actions I want to highlight are `woocommerce_blocks_loaded` and `plugins_loaded`.

When loading from the feature plugin, `woocommerce_blocks_loaded` happens before `plugins_loaded` but when loading from WC Core, `plugins_loaded` happens first.

This causes bugs when plugins wait for `plugins_loaded` before doing their initialisation, and part of that initialisation is doing some work on `woocommerce_blocks_loaded`.

This was most recently apparent in WooCommerce Min/Max Quantities but I am sure this could trip other developers up too.

## To reproduce

Add this code to a plugin's main file, outside of any classes/functions.

```php
add_action('plugins_loaded', function () {
    var_dump( 'plugins_loaded' );
});
add_action('woocommerce_blocks_loaded', function () {
    var_dump( 'woocommerce_blocks_loaded' );
});
```

- With the feature plugin active, notice the order of output.
- Deactivate the feature plugin and notice the order of the output.

## Expected behavior

I would expect actions to be fired in a consistent order regardless of how the plugin is loaded.

**Comment by opr** [comment]
cc @PanosSynetos

**Comment by mikejolley** [comment]
[As a plugin, this line initialises the package which in turn calls the action.](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/woocommerce-gutenberg-products-block.php#L153) Note this is done DURING `plugins_loaded`, so depending on what hooked in first will determine the order in which things run. Note it’s during, not after nor before `plugins_loaded`.

When blocks is loaded by WooCommerce core as a package, the main `woocommerce-gutenberg-products-block.php` file is never called. Instead, [Woo core loads our package on `plugins_loaded` here.](https://github.com/woocommerce/woocommerce/blob/c4628b966be16218a99d093cf3672ad79a4f235e/plugins/woocommerce/src/Packages.php#L37)

Both use plugins_loaded. I believe the issue is how the hook order is being observed. 

```
add_action('plugins_loaded', function () {
    var_dump( 'plugins_loaded' );
});
```

The above debugging code used in the OP runs on `plugins_loaded` too but they are ALL running on priority 10. This means the functions are called in whatever order they are defined, and usually, this comes down to plugin filename (alphabetically). We have no control over that aspect.

For plugins using `plugins_loaded`  they can mitigate the issue by using a later priority hook to ensure that all plugins, blocks too, are loaded in time. e.g. 

```
add_action( 'plugins_loaded','callback', 11 );
```

Or they can use `woocommerce_blocks_loaded`  to guarantee blocks have been initialised.

I don’t think there is much we can do here besides document in all honesty. This is how WordPress loads plugins - if you’re fighting for priority 10, there is always a chance of failure. I believe this to be a documentation issue. 

It's unfortunate there is some inconsistency, but it comes down to the plugin names—`woocommerce` vs `woocommerce-products-block`. `woocommerce` will always load before `woocommerce-products-block`, with other plugins somewhere inbetween.

**Comment by wavvves** [comment]
There is some confusion derived from the phrasing we're using, namely `woocommerce_blocks_loaded` vs `plugins_loaded/plugin_loaded`. 

What is loaded for WP? It means files were included, and the code is available.
Looking [here](https://github.com/WordPress/WordPress/blob/master/wp-settings.php#L443) (up until [L480](https://github.com/WordPress/WordPress/blob/master/wp-settings.php#L480)), we see the `plugin_loaded` and `plugins_loaded` getting triggered. The first one gets triggered whenever a plugin's main file is included. Once all `wp_get_active_and_valid_plugins()`  get loaded, WP then fires the `plugins_loaded` hook.

Now, we are instantiating our plugin during the `plugins_loaded`  call. That might occur sooner or later than other plugins being instantiated. But we're calling `do_action( 'woocommerce_blocks_loaded' );`  [on (practically after) our instantiation](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/src/Domain/Bootstrap.php#L78), when, to be accurate, "loaded" already occurred in terms of WP phrasing.

So to sum it up @opr 

- Hooking to `plugins_loaded`  we can only be 100% sure all of the code from blocks is available for usage, but not take for granted a usable plugin state.
- Hooking into `woocommerce_blocks_loaded` tells us Blocks is loaded and available for usage, but other plugins might not be.

None of the above can guarantee us all plugins are ready to be used in the instantiated form.

**Comment by ralucaStan** [comment]
Closing this ticket based on above comments. I've opened https://github.com/woocommerce/woocommerce-blocks/issues/7315 for documenting the `woocommerce_blocks_loaded` action and why extensions should not rely on `plugins_loaded`
